### ISSUE TITLE
AWS SSO integration

### ISSUE THREAD

**Issue by andreacavagna01** [Noovolari/leapp#38]
**Is your feature request related to a problem? Please describe.**
Add AWS SSO as an identity provider for AWS.

## **AWS SSO AUTH FLOW**

AWS SSO provides OAuth 2.0 Device Authorization Grant compatible APIs to enable command line tools to use AWS SSO to obtain AWS credentials. While those APIs are already documented and used by the AWS Command Line Interface version 2 (AWS CLI) they are not yet actively promoted by AWS.

The corresponding AWS APIs are:

- [AWS SSO OIDC API](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference)
- [AWS SSO Portal API](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference)

With Leapp, we have to setup the integration with this specific Identity provider.

The thought is to provide an **Integrations** page to manage every external integration with Leapp.
So in the menu we will find and Integration button.

![leapp menu](https://user-images.githubusercontent.com/17701428/97583767-9574ac00-19f7-11eb-9f86-27f33bf451dd.png)

By clicking the integrations menu you will see an info to register AWS SSO as identity provider.

![Tavola da disegno – 13](https://user-images.githubusercontent.com/17701428/97583811-a4f3f500-19f7-11eb-9fed-1f0fc522b633.png)

Whenever you click the AWS SSO button the auth flow must start.

The Auth flow in AWS SSO is made up with OIDC and is divided in three different steps:

1. [RegisterClient](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_RegisterClient.html)
2. [StartDeviceAuthorization](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html)
3. [CreateToken](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_CreateToken.html)

> CREDENTIALS ARE NOT NEEDED FOR THOSE THREE CALLS!

The first thing we have to do is to register a local client.

> The AWS SSO OIDC service currently implements only the portions of the OAuth 2.0 Device Authorization Grant standard ([https://tools.ietf.org/html/rfc8628](https://tools.ietf.org/html/rfc8628)) that are necessary to enable SSO authentication with the AWS CLI. Support for other OIDC flows frequently needed for native applications, such as Authorization Code Flow (+ PKCE), is planned in future releases.

## RegisterClient

As soon as the button is started, the RegisterClient Api will be invoked.

```json
aws sso-oidc register-client --client-name leapp --client-type public
```

> NOTES: public is the only client type available at the moment.

it will generate a response like the one below:

```json
{
    "clientId": "Clhq.....tMQ",
    "clientSecret": "eyJra.....wi...lK",
    "clientIdIssuedAt": 1603898686,
    "clientSecretExpiresAt": 1611674686
}
```

By decoding the Client secret you will see:

```json
{
  "expired": false,
  "clientId": {
    "value": "R-j5yN-4-TPNs...tMQ"
  },
  "clientName": "my-client",
  "clientType": "PUBLIC",
  "templateArn": null,
  "templateContext": null,
  "expirationTimestamp": 1594932567.612,
  "createdTimestamp": 1587156567.612,
  "updatedTimestamp": 1587156567.612,
  "createdBy": null,
  "updatedBy": null
}
```

We have to locally save those info and pass them into the next call 

## StartDeviceAuthorization

in order to make this call the user of Leapp need to have registered into AWS SSO and provide a valid start-url, that is the portal url

![url portal](https://user-images.githubusercontent.com/17701428/97583957-c654e100-19f7-11eb-8344-5188ee05d0ec.png)

Having a valid OIDC client registration we can now initiate the device authorization flow using the [StartDeviceAuthorization](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html) API action.

```jsx
aws sso-oidc start-device-authorization 
--client-id mVZBEoa-gUBj8nugHZUIsWV1LXdlc3QtMQ 
--client-secret eyJra...a_tV-htKetQH8 
--start-url ENDPOINTURL
```

That will produce a response like this: 

```json
{
  "deviceCode": "8Acq...DUg",
  "expiresIn": 600,
  "interval": 1,
  "userCode": "RPXP-JSQA",
  "verificationUri": "https://device.sso.eu-central-1.amazonaws.com/",
  "verificationUriComplete": "https://device.sso.eu-central-1.amazonaws.com/?user_code=RPXP-JSQA"
}
```

Where:

- **[deviceCode](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):** The short-lived code that is used by the device when polling for a session token.
- **[expiresIn](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):** Indicates the number of seconds in which the verification code will become invalid.
- **[interval](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):** Indicates the number of seconds the client must wait between attempts when polling for a session.
- **[userCode](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):** A one-time user verification code. This is needed to authorize an in-use device.
- **[verificationUri](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):** The URI of the verification page that takes the `userCode` to authorize the device.
- **[verificationUriComplete](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_StartDeviceAuthorization.html#API_StartDeviceAuthorization_ResponseSyntax):**An alternate URL that the client can use to automatically launch a browser. This process skips the manual step in which the user visits the verification page and enters their code.

> In order to verify this device, **Leapp** will prompt you to the LOGIN PAGE and the verification code page, by opening a web browser with the `verificationUriComplete`

After the verification of the device we can now call the `CreateToken` API

## CreateToken

Having the `userCode`and `deviceCode`values we can now us the [CreateToken](https://docs.aws.amazon.com/singlesignon/latest/OIDCAPIReference/API_CreateToken.html) API action to obtain a device access token. However, before we can request any tokens we need to open the `verificationUriComplete` URL in a web browser, complete authentication and provide authorization. At this point the end-user has to provide personal credentials in order to authenticate himself, for example by entering his username, password and maybe also has to provide multi factory authentication (MFA) credentials.

```jsx
curl -X POST https://oidc.eu-central-1.amazonaws.com/token -d 
'{"clientId": "R-j5yN-4-TPNs...tMQ", "clientSecret": "eyJr...x74", "deviceCode": "8Acq...DUg", "grantType": "urn:ietf:params:oauth:grant-type:device_code"}'
```

If everything worked out as expected, the token endpoint will return a valid accessToken, which we later can use with the **AWS SSO Portal API** to obtain temporary AWS credentials using the GetRoleCredentials action.

**The returned tokens are valid for 8 hours.**

> This API call is region sensible, so we have to extract the region from the `verificationUriComplete` url with a RegEx!

With this token we can now call the [AWS SSO Portal API](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference) to autogenerate sessions in Leapp.

# AWS SSO Portal API

In order to fill the session available with AWS SSO we have to use those APIs

The following actions are supported:

- [GetRoleCredentials](https://docs.aws.amazon.com/it_it/singlesignon/latest/PortalAPIReference/API_GetRoleCredentials.html)
- [ListAccountRoles](https://docs.aws.amazon.com/it_it/singlesignon/latest/PortalAPIReference/API_ListAccountRoles.html)
- [ListAccounts](https://docs.aws.amazon.com/it_it/singlesignon/latest/PortalAPIReference/API_ListAccounts.html)
- [Logout](https://docs.aws.amazon.com/it_it/singlesignon/latest/PortalAPIReference/API_Logout.html)

and in order to complete this action we need scan for account and roles inside them, so it's a time consuming moment.

so a loading modal will appear:

![Tavola da disegno – 15](https://user-images.githubusercontent.com/17701428/97584098-f308f880-19f7-11eb-87d3-02d0275927a9.png)

What happens whenever this modal will appear?

## ListAccounts

Firstly we have to check all the accounts available for the current user.

Lists all AWS accounts assigned to the user. These AWS accounts are assigned by the administrator of the account.

```bash
GET /assignment/accounts?max_result=maxResults&next_token=nextToken HTTP/1.1
x-amz-sso_bearer_token: accessToken
```

- **[accessToken](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_ListAccounts.html#API_ListAccounts_RequestSyntax):** The token issued by the `CreateToken` API call. **Required: Yes**
- **[maxResults](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_ListAccounts.html#API_ListAccounts_RequestSyntax):** This is the number of items clients can request per page. **Valid Range: Minimum value of 1. The maximum value of 100.**
- **[nextToken](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_ListAccounts.html#API_ListAccounts_RequestSyntax):**  When requesting subsequent pages, this is the page token from the previous response output. **Required: NO**

**RESPONSE**

```json
{
   "accountList": [ 
      { 
         "accountId": "string",
         "accountName": "string",
         "emailAddress": "string"
      }
   ],
   "nextToken": "string"
}
```

The following data is returned in JSON format by the service.

- **[accountList](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_ListAccounts.html#API_ListAccounts_ResponseSyntax):** A paginated response with the list of account information and the next token if more results are available.
    - **accountId:** The identifier of the AWS account that is assigned to the user.
    - **accountName**The display name of the AWS account that is assigned to the user.
    - **emailAddress**The email address of the AWS account that is assigned to the user. Length

For each account retrieved by this API, we need to check all the possible roles inside it.

## [ListAccountRoles](https://docs.aws.amazon.com/it_it/singlesignon/latest/PortalAPIReference/API_ListAccountRoles.html)

```json
aws sso list-account-roles --access-token eyJlbmMiOiJBM...nZEpteo-dHw --account-id 198863347786 
```

Response: 

```json
{
    "roleList": [
        {
            "roleName": "ViewOnlyAccess",
            "accountId": "198000000086"
        }
    ]
}
```

After that for each pair of Account and role found we will add an AWS session that is available to get the correct credentials.

![Tavola da disegno – 7](https://user-images.githubusercontent.com/17701428/97584807-b689cc80-19f8-11eb-8b20-96a577856806.png)

in order to retrieve credentials we have to  make this call

## GetRoleCredentials

Having a valid access token, we can use the [AWS SSO Portal API](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference) to obtain temporary AWS credentials using the [GetRoleCredentials](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_GetRoleCredentials.html) action.

```bash
curl 'https://portal.sso.eu-central-1.amazonaws.com/federation/credentials?account_id=999999999999&role_name=MyIamRoleName' -H 'x-amz-sso_bearer_token: eyJl...Blw'
```

Which will then provide us with temporary AWS session credentials, which by default are valid for 12 hours.

```json
{
  "roleCredentials": {
    "accessKeyId": "ASIA...Z3XM",
    "secretAccessKey": "+t6UnVLWia...hFOzGL",
    "sessionToken": "IQoJb3JpZ2lu...dK6",
    "expiration": 150000008000
  }
}
```

I tried several additional query parameters (`session_duration`, `duration_seconds`, …) to figure out whether one can choose a shorter session duration, but it looks like 12 hours is all you can get. While [AWS STS API](https://docs.aws.amazon.com/STS/latest/APIReference/Welcome.html)’s actions like [AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) or [AssumeRoleWithWebIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) allow callers to specify custom session duration, [GetRoleCredentials](https://docs.aws.amazon.com/singlesignon/latest/PortalAPIReference/API_GetRoleCredentials.html) does not yet support that.

# Final thoughts

there are still two actions to be implemented in the App, the `Logout` action and the `Sync` action.

The first call the api call of the portal

The second one will synchronize the session with the current configuration.

**Comment by benkehoe** [comment]
You need AWS region as an input in the setup step in addition to the SSO start URL. This is the region that the AWS SSO instance is deployed in, not the region that a given config profile will use for using the retrieved credentials. Note the `eu-central-1` in `https://portal.sso.eu-central-1.amazonaws.com/federation/credentials?account_id=999999999999&role_name=MyIamRoleName`

**Comment by andreacavagna01** [comment]
I noted that inside the `"verificationUriComplete` response there is the info about the region.

example:  "https://device.sso.eu-central-1.amazonaws.com/?user_code=ABCD-EDFR"

So I thought to take the region with a split string from that with code, could it work?

**Comment by urz9999** [comment]
either way i think

**Comment by benkehoe** [comment]
I believe the region is required for the sso-oidc API as well, so I don't think you can get to that point without a region.

**Comment by andreacavagna01** [comment]
So it needs to be added in the UI with the Portal Url; I Will add it. 

What's your idea of the UX/UI flow?

**Comment by andreacavagna01** [comment]
Released in v.0.4.0
